Run integration tests with runtime

vatsrahul1001 opened the pull request:
# Description
## What is the current behavior?

Currently we have a static [deployment](https://e2etesting.astronomer.run/dmmkey8w/home) in which we execute 
[master_dag](https://github.com/astronomer/astro-sdk/blob/main/python-sdk/tests_integration/astro_deploy/master_dag.py), however, we do not have a way to run these tests with every new runtime alpha release before RC


## What is the new behavior?
This PR has below refactoring to support the execution of master-dag with every runtime alpha release() from turbulence repo 

1.  In example dags I noticed we were using some duplicate connections.
2. Refactoring of master_dag with respect to execution with every runtime release


Below are the steps that turbulence performs to create a airflow deployment.
    1. Make a new deployment on a cloud stage environment with GraphQL calls.
    2. Creates a dockerfile with the latest runtime image(alpha, stable),latest astro-sdk, example dags and we build this dockerfile and create a image hash.
    4. Use Astro CLI to update the newly created deployment(in # 1) with the image has created (in # 2)





## Does this introduce a breaking change?
No, this is a new feature and does not affect existing functionality.

tatiana commented on python-sdk/example_dags/example_bigquery_dynamic_map_task.py:
```diff
@@ -18,9 +18,15 @@
 
 ASTRO_BIGQUERY_DATASET = os.getenv("ASTRO_BIGQUERY_DATASET", "dag_authoring")
-ASTRO_GCP_CONN_ID = os.getenv("ASTRO_GCP_CONN_ID", "google_cloud_default")
+ASTRO_GCP_CONN_ID = os.getenv("ASTRO_GCP_CONN_ID", "google_cloud_platform")
```
Why are we moving out from the `google_cloud_default` default...? If we need to this to have another value due to the Astro deployment, we can always define the environment variable `ASTRO_GCP_CONN_ID` in those scenarios.

tatiana commented on python-sdk/example_dags/example_google_bigquery_gcs_load_and_save.py:
```diff
@@ -22,19 +22,29 @@
         ),
-        output_table=Table(name="imdb_movies", conn_id="bigquery", metadata=Metadata(schema="astro")),
+        output_table=Table(
+            name="imdb_movies", conn_id="google_cloud_platform", metadata=Metadata(schema="astro")
```
Let's not hard-code the connection ID here, but define it on the top of the file:
```
ASTRO_GCP_CONN_ID = os.getenv("ASTRO_GCP_CONN_ID", "google_cloud_default")
```

The [same comment](https://github.com/astronomer/astro-sdk/pull/1937/files#r1198689257) about the connection ID applies here.

tatiana commented on python-sdk/example_dags/example_google_bigquery_gcs_load_and_save.py:
```diff
@@ -58,7 +68,7 @@ def extract_top_5_movies(input_df: pd.DataFrame):
         output_file=File(
             path=f"{gcs_bucket}/{{{{ task_instance_key_str }}}}/all_movies.csv",
-            conn_id="gcp_conn",
+            conn_id="google_cloud_platform",
```
```suggestion
            conn_id= ASTRO_GCP_CONN_ID,
```

tatiana commented on python-sdk/example_dags/example_google_bigquery_gcs_load_and_save.py:
```diff
@@ -70,7 +80,7 @@ def extract_top_5_movies(input_df: pd.DataFrame):
         output_file=File(
             path=f"{gcs_bucket}/{{{{ task_instance_key_str }}}}/top_5_movies.csv",
-            conn_id="gcp_conn",
+            conn_id="google_cloud_platform",
```
```suggestion
            conn_id= ASTRO_GCP_CONN_ID,
```

vatsrahul1001 replied:
Resolved

tatiana commented on python-sdk/example_dags/example_google_bigquery_gcs_load_and_save.py:
```diff
@@ -22,19 +22,29 @@
         ),
-        output_table=Table(name="imdb_movies", conn_id="bigquery", metadata=Metadata(schema="astro")),
+        output_table=Table(
+            name="imdb_movies", conn_id="google_cloud_platform", metadata=Metadata(schema="astro")
```
```suggestion
            name="imdb_movies", conn_id=ASTRO_GCP_CONN_ID, metadata=Metadata(schema="astro")
```

vatsrahul1001 replied:
Resolved

tatiana commented on python-sdk/example_dags/example_load_file.py:
```diff
@@ -117,15 +117,15 @@
     aql.load_file(
         input_file=File("s3://tmp9/homes_main.csv", conn_id=AWS_CONN_ID),
-        output_table=Table(conn_id="bigquery", metadata=Metadata(schema="astro")),
+        output_table=Table(conn_id="google_cloud_platform", metadata=Metadata(schema="astro")),
```
```suggestion
        output_table=Table(conn_id=ASTRO_GCP_CONN_ID, metadata=Metadata(schema="astro")),
```

vatsrahul1001 replied:
resolved

tatiana commented on python-sdk/example_dags/example_load_file.py:
```diff
@@ -117,15 +117,15 @@
     aql.load_file(
         input_file=File("s3://tmp9/homes_main.csv", conn_id=AWS_CONN_ID),
-        output_table=Table(conn_id="bigquery", metadata=Metadata(schema="astro")),
+        output_table=Table(conn_id="google_cloud_platform", metadata=Metadata(schema="astro")),
```
```suggestion
        output_table=Table(conn_id=ASTRO_GCP_CONN_ID, metadata=Metadata(schema="astro")),
```

tatiana commented on python-sdk/example_dags/example_load_file.py:
```diff
@@ -160,7 +160,7 @@
     aql.load_file(
         input_file=File("s3://astro-sdk/sample_pattern", conn_id=AWS_CONN_ID, filetype=FileType.CSV),
-        output_table=Table(conn_id="bigquery", metadata=Metadata(schema="astro")),
+        output_table=Table(conn_id="google_cloud_platform", metadata=Metadata(schema="astro")),
```
```suggestion
        output_table=Table(conn_id=ASTRO_GCP_CONN_ID, metadata=Metadata(schema="astro")),
```

tatiana commented on python-sdk/example_dags/example_load_file.py:
```diff
@@ -169,10 +169,10 @@
             filetype=FileType.CSV,
         ),
-        output_table=Table(conn_id="bigquery", metadata=Metadata(schema="astro")),
+        output_table=Table(conn_id="google_cloud_platform", metadata=Metadata(schema="astro")),
```
```suggestion
        output_table=Table(conn_id=ASTRO_GCP_CONN_ID, metadata=Metadata(schema="astro")),
```

tatiana commented on python-sdk/example_dags/example_load_file.py:
```diff
@@ -193,7 +193,7 @@
         input_file=File(
             "gs://astro-sdk/workspace/sample_pattern.csv",
-            conn_id="bigquery",
+            conn_id="google_cloud_platform",
```
```suggestion
            conn_id=ASTRO_GCP_CONN_ID,
```

tatiana commented on python-sdk/example_dags/example_load_file.py:
```diff
@@ -227,10 +227,10 @@
         input_file=File(
             "gs://astro-sdk/workspace/sample_pattern.csv",
-            conn_id="bigquery",
+            conn_id="google_cloud_platform",
```
```suggestion
            conn_id=ASTRO_GCP_CONN_ID,
```

tatiana commented on python-sdk/example_dags/example_load_file.py:
```diff
@@ -227,10 +227,10 @@
             filetype=FileType.CSV,
         ),
-        output_table=Table(conn_id="bigquery", metadata=Metadata(schema="astro")),
+        output_table=Table(conn_id="google_cloud_platform", metadata=Metadata(schema="astro")),
```
```suggestion
        output_table=Table(conn_id=ASTRO_GCP_CONN_ID, metadata=Metadata(schema="astro")),
```

tatiana commented on python-sdk/example_dags/example_merge.py:
```diff
@@ -43,7 +44,7 @@ def my_df_func(input_df: DataFrame):
         input_file=File(DATA_DIR + "sample.csv"),
         output_table=Table(
-            conn_id="bigquery",
+            conn_id="google_cloud_platform",
```
```suggestion
            conn_id=ASTRO_GCP_CONN_ID,
```

tatiana commented on python-sdk/example_dags/example_merge.py:
```diff
@@ -55,7 +56,7 @@ def my_df_func(input_df: DataFrame):
         input_file=File(DATA_DIR + "sample.csv"),
         output_table=Table(
-            conn_id="bigquery",
+            conn_id="google_cloud_platform",
```
```suggestion
            conn_id=ASTRO_GCP_CONN_ID,
```

tatiana commented on python-sdk/example_dags/example_merge.py:
```diff
@@ -66,7 +67,7 @@ def my_df_func(input_df: DataFrame):
         input_file=File(DATA_DIR + "sample_part2.csv"),
         output_table=Table(
-            conn_id="bigquery",
+            conn_id="google_cloud_platform",
```
```suggestion
            conn_id=ASTRO_GCP_CONN_ID,
```

tatiana commented on python-sdk/example_dags/example_mssql_transform.py:
```diff
@@ -56,24 +63,26 @@ def union_table_and_dataframe(input_table: Table, input_dataframe: pd.DataFrame)
         input_file=File(path="s3://astro-sdk/imdb_v2.csv"),
         task_id="load_csv",
-        output_table=Table(conn_id="mssql_conn"),
+        output_table=Table(conn_id="mysql_conn"),
```
Let's define a constant on the top of the file so we avoid hard-code constants in the middle of the DAG

vatsrahul1001 replied:
resolved

tatiana commented on python-sdk/example_dags/example_mssql_transform.py:
```diff
@@ -56,24 +63,26 @@ def union_table_and_dataframe(input_table: Table, input_dataframe: pd.DataFrame)
         input_table=imdb_movies,
         output_table=Table(
-            conn_id="mssql_conn",
+            conn_id="mysql_conn",
```
Same as https://github.com/astronomer/astro-sdk/pull/1937/files#r1198698859

tatiana commented on python-sdk/example_dags/example_mssql_transform.py:
```diff
@@ -56,24 +63,26 @@ def union_table_and_dataframe(input_table: Table, input_dataframe: pd.DataFrame)
         input_table=imdb_movies,
         output_table=Table(
-            conn_id="mssql_conn",
+            conn_id="mysql_conn",
```
Same as https://github.com/astronomer/astro-sdk/pull/1937/files#r1198698859

tatiana commented on python-sdk/tests_integration/astro_deploy/astro_sdk_connections.py:
```diff
@@ -0,0 +1,87 @@
+
+
+@task
+def replace_env_var_in_yaml():
+    with open(connections_yaml) as file:
+        yaml_content = yaml.safe_load(file)
+
+        # Recursively replace environment variables in the YAML content
+    connections_details = recursive_replace_env_vars(yaml_content)
+
+    # Save the modified YAML content back to the file
+    with open(connections_yaml, "w") as file:
+        yaml.dump(connections_details, file, default_flow_style=False)
+
+
+def recursive_replace_env_vars(data):
+    if isinstance(data, dict):
+        for key, value in data.items():
+            if isinstance(value, str):
+                # Check if the string value is an environment variable
+                if value.startswith("$") and value[1:] in os.environ:
+                    data[key] = os.environ[value[1:]]
+            else:
+                # Recursively replace environment variables in nested data structures
+                data[key] = recursive_replace_env_vars(value)
+    elif isinstance(data, list):
+        # Recursively replace environment variables in lists
+        data = [recursive_replace_env_vars(item) for item in data]
+
+    return data
+
```
We may not need to define all this custom code. We could use the Python standard library: `os.path.expandvars`, as documented in https://docs.python.org/3/library/os.path.html#os.path.expandvars.

You can see an example of how we accomplished this in the Python SDK code:
https://github.com/astronomer/astro-sdk/blob/2627e67c47506601e72e7d53f1d515adce0b317a/python-sdk/tests_integration/conftest.py#L46-L68

utkarsharma2 replied:
+1

vatsrahul1001 replied:
Removed this change, will be using secrets backend for connection storage

tatiana commented on python-sdk/tests_integration/astro_deploy/astro_sdk_connections.py:
```diff
@@ -0,0 +1,87 @@
+            session.add(conn)
+
+
+@dag(
+    dag_id="create_sdk_required_connections",
+    start_date=datetime(1970, 1, 1),
+    schedule_interval=None,
+    tags=["setup"],
+    doc_md=dedent(
+        """
+        This DAG creates connections required by astro sdk dags connection
+        This Dag takes as input connections_template.yaml(connections information with some unresolved Env vars),
+        Resolve the variables in file and create connections in airflow, we can add new connections in connections_template.yaml
+        to push them to airflow
+
+        Run this before you run any dags that depend on that connection
+        """
+    ),
+)
+def setup():
+    replace_env_var_in_yaml()
+    create_connections()
```
I don't feel comfortable with a DAG setting connections in our deployment - DAGs should not change infrastructure settings, such as connections.

utkarsharma2 replied:
Maybe we can add a run Python script as part of CI Job to sync connections from SDK repo. Code need to be changed for this, since session is no longer useful now. We can maybe use - https://airflow.apache.org/docs/apache-airflow/stable/stable-rest-api-ref.html#operation/post_connection API to populate airfclow connections.

vatsrahul1001 replied:
Removed this change, will be using secrets backend for connection storage

tatiana commented on python-sdk/tests_integration/astro_deploy/master_dag.py:
```diff
@@ -18,15 +18,21 @@
 SLACK_WEBHOOK_CONN = os.getenv("SLACK_WEBHOOK_CONN", "http_slack")
 SLACK_USERNAME = os.getenv("SLACK_USERNAME", "airflow_app")
 
+is_runtime_release = Variable.get("IS_RUNTIME_RELEASE", default_var="False")
```
Could this be an Airflow configuration as opposed to a `Variable`?

The advantage of using config is that in line 32, we'd be able to use `getboolean`:
https://github.com/apache/airflow/blob/54f1fb0574a6ecf8f415bbf6da1aaf6f1999bb29/airflow/configuration.py#L765

vatsrahul1001 replied:
Not able to add a new config option however converted to bool value after taking srt value from env

tatiana commented on python-sdk/tests_integration/astro_deploy/master_dag.py:
```diff
@@ -18,15 +18,21 @@
 
         airflow_version = context["ti"].xcom_pull(task_ids="get_airflow_version")
-        airflow_version_message = f"Airflow version for the below astro-sdk run is `{airflow_version}` \n\n"
+        if is_runtime_release == "TRUE":
```
This seems suboptimal, why not using something like `getboolean`?
https://github.com/apache/airflow/blob/54f1fb0574a6ecf8f415bbf6da1aaf6f1999bb29/airflow/configuration.py#L765

utkarsharma2 commented on python-sdk/tests_integration/astro_deploy/astro_sdk_connections.py:
```diff
@@ -0,0 +1,87 @@
+from airflow.utils.db import create_default_connections
+from airflow.utils.session import create_session
+
+connections_yaml = f"{os.environ['AIRFLOW_HOME']}/include/connections_template.yaml"
```
@vatsrahul1001 How are we getting this file?

vatsrahul1001 replied:
Removed this change, will be using secrets backend for connection storage

utkarsharma2 commented on python-sdk/tests_integration/astro_deploy/astro_sdk_connections.py:
```diff
@@ -0,0 +1,87 @@
+@task
+def create_connections():
+    with open(connections_yaml) as fp:
+        yaml_with_env = os.path.expandvars(fp.read())
```
In reference to above comment - https://github.com/astronomer/astro-sdk/pull/1937/files#r1198702725
`yaml_with_env = os.path.expandvars(fp.read())` <- this didn't do the trick for you? we may not need `replace_env_var_in_yam`l task as all

vatsrahul1001 replied:
Removed this change, will be using secrets backend for connection storage

utkarsharma2 commented on python-sdk/tests_integration/astro_deploy/connections_template.yaml:
```diff
@@ -0,0 +1,85 @@
+connections:
```
+1

vatsrahul1001 replied:
Removed this change, will be using secrets backend for connection storage

tatiana commented on python-sdk/example_dags/calculate_popular_movies.py:
```diff
@@ -20,10 +20,7 @@ def top_five_animations(input_table: Table):
-    schedule_interval=None,
-    start_date=datetime(2000, 1, 1),
-    catchup=False,
+    "calculate_popular_movies", schedule_interval=None, start_date=datetime(2000, 1, 1), catchup=False
```
Why did we change this line...?

vatsrahul1001 replied:
will remove formatting

tatiana commented on python-sdk/example_dags/example_check_table.py:
```diff
@@ -7,10 +7,7 @@
-    schedule_interval=None,
-    start_date=datetime(2000, 1, 1),
-    catchup=False,
+    "data_validation_check_table", schedule_interval=None, start_date=datetime(2000, 1, 1), catchup=False
```
Why did we change this line...?

vatsrahul1001 replied:
will remove formatting

tatiana commented on python-sdk/example_dags/example_google_bigquery_gcs_load_and_save.py:
```diff
@@ -22,19 +22,28 @@
 from astro.files import File
 from astro.table import Metadata, Table
 
+ASTRO_GCP_CONN_ID = os.getenv("ASTRO_GCP_CONN_ID", "google_cloud_default")
```
Always good to see constants extracted to the code and put on top of the file :)

tatiana commented on python-sdk/example_dags/example_load_file.py:
```diff
@@ -352,7 +357,7 @@
         input_file=File("s3://tmp9/homes_main.csv", conn_id=AWS_CONN_ID),
         output_table=Table(
-            conn_id=MSSQL_CONN_ID,
+            conn_id=MYSQL_CONN_ID,
```
Why did we replace the `MSSQL_CONN_ID`(MS SQL)  by a MySQL one?

tatiana replied:
@vatsrahul1001 the names are very similar, but MySQL != MSSQL.

MySQL is an open-source relational database, while MSSQL Server is a Microsoft-developed db.

vatsrahul1001 replied:
@tatiana i have updated MSSQL

tatiana commented on python-sdk/example_dags/example_mssql_transform.py:
```diff
@@ -56,24 +64,25 @@ def union_table_and_dataframe(input_table: Table, input_dataframe: pd.DataFrame)
         input_file=File(path="s3://astro-sdk/imdb_v2.csv"),
         task_id="load_csv",
-        output_table=Table(conn_id="mssql_conn"),
+        output_table=Table(conn_id=ASTRO_MYSQL_CONN_ID),
```
Why did we replace the MS SQL connection with a MySQL one?

vatsrahul1001 replied:
@tatiana i see there were two conn one with MSSQL_CONN_ID and other MYSQL_CONN_ID, so I thought to use one

tatiana replied:
They are different databases, as mentioned in https://github.com/astronomer/astro-sdk/pull/1937/files/3d38452bdab381179d900467a276824c87913c1f#r1211368378

vatsrahul1001 replied:
@tatiana i have updated MSSQL

tatiana commented on python-sdk/example_dags/example_transform_file.py:
```diff
@@ -10,12 +10,7 @@
-    start_date=START_DATE,
-    catchup=False,
-) as dag:
+with DAG("example_transform_file", schedule_interval=None, start_date=START_DATE, catchup=False) as dag:
```
Why did this become a one-liner?

vatsrahul1001 replied:
@will remove the formating

tatiana commented on python-sdk/example_dags/example_amazon_s3_snowflake_transform.py:
```diff
@@ -43,8 +43,8 @@ def aggregate_data(df: pd.DataFrame):
         "email_on_failure": False,
-        "retries": 0,
-        "retry_delay": timedelta(minutes=5),
+        "retries": 1,
```
Why did we change 12 DAGs to have `"retries": 1`?

vatsrahul1001 replied:
@tatiana Earlier these DAG did not have retries before added to give one more retry in case of failure

tatiana commented on python-sdk/tests_integration/astro_deploy/master_dag.py:
```diff
@@ -17,16 +17,21 @@
 SLACK_CHANNEL = os.getenv("SLACK_CHANNEL", "#provider-alert")
 SLACK_WEBHOOK_CONN = os.getenv("SLACK_WEBHOOK_CONN", "http_slack")
 SLACK_USERNAME = os.getenv("SLACK_USERNAME", "airflow_app")
+IS_RUNTIME_RELEASE = bool(os.getenv("IS_RUNTIME_RELEASE", False))
```
Do we need to introduction of a new configuration? Could we infer if there is a runtime image just based on the environment variable `ASTRONOMER_RUNTIME_VERSION`?

vatsrahul1001 replied:
@tatiana Reason for adding Env Var is to set the schedule of a master DAG  NONE when this DAG is triggered by automation framework(when a new runtime is released), I am adding IS_RUNTIME_RELEASE in astro cloud automation deployment, using ENV ASTRONOMER_RUNTIME_VERSION will not help as this var is in every runtime deployment

tatiana replied:
Ah, I see, thank you very much, @vatsrahul1001 !

tatiana commented on python-sdk/tests_integration/astro_deploy/master_dag.py:
```diff
@@ -17,16 +17,21 @@
+        if IS_RUNTIME_RELEASE:
+            airflow_version_message = f"Results generated for latest Runtime version {os.environ['ASTRONOMER_RUNTIME_VERSION']} with {os.environ['AIRFLOW__CORE__EXECUTOR']}  \n\n"
+        else:
+            airflow_version_message = f"Airflow version for the below astro-sdk run is `{airflow_version}` with {os.environ['AIRFLOW__CORE__EXECUTOR']} \n\n"
```
How is Airflow installed in this scenario? I'm not familiar with the Turbulence approach, and since it is a private repo, it would be worth giving meaningful information as part of this PR.

Do we use another docker container, built with the latest version of Airflow. What is this latest version of Airflow (nightly build, main..)?

vatsrahul1001 replied:
@tatiana Below are the steps that turbulence performs to create a airflow deployment.

1. Make a new deployment on a cloud stage environment with GraphQL calls.
2. Creates a dockerfile with the latest runtime image(alpha, stable),latest astro-sdk, example dags and we build this dockerfile and create a image hash.
3. Use Astro CLI to update the newly created deployment(in #1) with the image has created (in #2)

tatiana replied:
Thanks, @vatsrahul1001 ! Is the latest Astro SDk the one in `main` or the last released Astro SDK?

vatsrahul1001 replied:
@tatiana we are using last released

vatsrahul1001 commented:
> @vatsrahul1001 we're super close to having this PR merged. I added some last few comments.
> 
> Please, I'd love to see the following addressed (I added some more details to some of these inline):
> 
> 1. Could you update the PR description so it is up-to-date with the current contents of this PR (I don't think "3. Created a new DAG, which will create all the connections that are necessary to execute the example dags" applies anymore. <RV>:Done
> 2. If we could avoid reference to a private repository (Turbulence) and describe as part of the repo (either documentation or comments in the code) how we're using the latest version of Airflow and what it is, it would be amazing.
<RV>:Update the PR desc
> 3. Why did we change so many files for retries?
<RV> Earlier these DAG did not have retries before added to give one more retry in case of failure
> 4. there seem to be some formatting-related changes, which should not be necessary. Ideally, we focus on the goal of the PR. 
<RV> Remove black formatting
> 5. It seems some MySQL connections mistakenly replaced some MSSQL connections.
<RV>: We were using  two conn one with MSSQL_CONN_ID and other MYSQL_CONN_ID and both connections were same, updated to use one of them

tatiana reviewed (commented):
I think we can merge this today, @vatsrahul1001 ! Just need to fix the MSSQL + MySQL and I had one last question - thanks a lot!

vatsrahul1001 commented:
> I think we can merge this today, @vatsrahul1001 ! Just need to fix the MSSQL + MySQL and I had one last question - thanks a lot!

@tatiana i have updated MSSQL

tatiana commented:
@vatsrahul1001, please, address the failing checks:
* pre-commit
* DeepSource
* broken test:
```
FAILED tests_integration/test_example_dags.py::test_example_dag[example_duckdb_load_transform_dataframe_and_save] - ValueError: You need to provide a table or a connection id
```

tatiana reviewed (commented):
@vatsrahul1001, please, address the failing checks:
* pre-commit
* DeepSource
* broken test:
```
FAILED tests_integration/test_example_dags.py::test_example_dag[example_duckdb_load_transform_dataframe_and_save] - ValueError: You need to provide a table or a connection id
```

vatsrahul1001 commented:
> @vatsrahul1001, please, address the failing checks:
> 
> * pre-commit
> * DeepSource
> * broken test:
> 
> ```
> FAILED tests_integration/test_example_dags.py::test_example_dag[example_duckdb_load_transform_dataframe_and_save] - ValueError: You need to provide a table or a connection id
> ```

Fixed all checks
